packages/node/src/login/oidc/refresh/TokenRefresher.ts

```diff
@@ -42,21 +50,65 @@ export interface ITokenRefresher {
 export default class TokenRefresher implements ITokenRefresher {
   constructor(
     @inject("storageUtility") private storageUtility: IStorageUtility,
-    @inject("tokenRequester") private tokenRequester: ITokenRequester
+    @inject("issuerConfigFetcher")
+    private issuerConfigFetcher: IIssuerConfigFetcher,
+    @inject("clientRegistrar") private clientRegistrar: IClientRegistrar
   ) {}
 
-  async refresh(localUserId: string): Promise<void> {
-    // Get the refresh token and the issuer
-    const refreshToken = await this.storageUtility.getForUser(
-      localUserId,
-      "refreshToken",
-      { errorIfNull: true, secure: true }
+  async refresh(
+    sessionId: string,
+    refreshToken?: string,
+    dpopKey?: JWK.ECKey
+  ): Promise<TokenSet> {
+    const oidcContext = await loadOidcContextFromStorage(
+      sessionId,
+      this.storageUtility,
+      this.issuerConfigFetcher
     );
-    /* eslint-disable camelcase */
-    await this.tokenRequester.request(localUserId, {
-      grant_type: "refresh_token",
-      refresh_token: refreshToken as string, // This cast can be safely made because getForUser will error if refreshToken is null
+    const issuer = new Issuer(configToIssuerMetadata(oidcContext.issuerConfig));
+    // This should also retrieve the client from storage
+    const clientInfo: IClient = await this.clientRegistrar.getClient(
+      { sessionId },
+      oidcContext.issuerConfig
+    );
+    const client = new issuer.Client({
+      client_id: clientInfo.clientId,
+      client_secret: clientInfo.clientSecret,
+    });
+
+    if (refreshToken === undefined) {
+      // TODO: in a next PR, look up storage for a refresh token
+      throw new Error(
+        `Missing a refresh token to refresh the access token associated to the session [${sessionId}]`
+      );
+    }
+
+    if (oidcContext.dpop && dpopKey === undefined) {
+      throw new Error(
+        `The key bound to the access token associated to the session [${sessionId}] must be provided to refresh it.`
+      );
+    }
+
+    const tokenSet = await client.refresh(refreshToken, {
+      DPoP: dpopKey?.toJWK(true),
     });
-    /* eslint-enable camelcase */
+
+    if (
+      tokenSet.access_token === undefined ||
+      tokenSet.id_token === undefined
+    ) {
+      // The error message is left minimal on purpose not to leak the tokens.
+      throw new Error("The IdP did not return the expected tokens.");
```

Comment:
From earlier comment, you do have the info to make this message much more helpful, I would add the missing info:
```suggestion
      const bothMissing: boolean = (tokenSet.access_token === undefined && tokenSet.id_token === undefined);
      const missingList = `${bothMissing ? "'access_token' and 'id_token'" : ${tokenSet.access_token === undefined ? 'access_token' : 'id_token'}`;
      throw new Error("The Identity Provider [${issuer}] did not return the expected token[${bothMissing ? 's' : ''} [${missingList}].");
```

Reply:
The test is removed in 8818797c031a7e51ec3c56e1bf2e195d54ff6dee, since actually only the access_token should be refreshed.